From 64bafab5ea258d3fe95d287a9d127ecce2ee8442 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 12 Sep 2025 11:28:39 +0200
Subject: [PATCH] drivers: nvmem: various fixes

Fix various error checks and allocations across the driver.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I9b2d4d06cd06f274ac4db88432a9abcc37791a74
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/489407
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/nvmem/stm32-tamp-nvram.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

--- a/drivers/nvmem/stm32-tamp-nvram.c
+++ b/drivers/nvmem/stm32-tamp-nvram.c
@@ -321,7 +321,9 @@ static int stm32_tamp_nvram_write_byte(s
 	if (!stm32_tamp_nvram_rights(priv, reg_idx, false))
 		return -EIO;
 
-	regmap_read(priv->bkpregs_regmap, offset_aligned, &read_value);
+	if (regmap_read(priv->bkpregs_regmap, offset_aligned, &read_value))
+		return -EIO;
+
 	to_be_writen_value = read_value & ~(0xFFUL << byte_in_word * 8);
 	to_be_writen_value |=  (u32)byte << (byte_in_word * 8);
 
@@ -338,7 +340,9 @@ static int stm32_tamp_nvram_read_byte(st
 	if (!stm32_tamp_nvram_rights(priv, reg_idx, true))
 		return -EIO;
 
-	regmap_read(priv->bkpregs_regmap, offset_aligned, &read_value);
+	if (regmap_read(priv->bkpregs_regmap, offset_aligned, &read_value))
+		return -EIO;
+
 	*byte = (read_value >> (byte_in_word * 8)) & 0xFF;
 
 	return 0;
@@ -363,8 +367,8 @@ static int stm32_tamp_nvram_read(void *c
 				dev_dbg(dev, "Backup register %u is not allowed to be read\n",
 					reg_idx);
 				temp_u32 = 0;
-			} else {
-				regmap_read(priv->bkpregs_regmap, i, &temp_u32);
+			} else if (regmap_read(priv->bkpregs_regmap, i, &temp_u32)) {
+				return -EIO;
 			}
 			memcpy(buf_u8, &temp_u32, sizeof(u32));
 			buf_u8 += sizeof(u32);
@@ -401,7 +405,8 @@ static int stm32_tamp_nvram_write(void *
 		if (i + sizeof(u32) <= total && IS_ALIGNED(i, sizeof(u32))) {
 			if (stm32_tamp_nvram_rights(priv, reg_idx, false)) {
 				memcpy(&temp_u32, buf_u8, sizeof(u32));
-				regmap_write(priv->bkpregs_regmap, i, temp_u32);
+				if (regmap_write(priv->bkpregs_regmap, i, temp_u32))
+					return -EIO;
 			} else {
 				dev_dbg(dev, "Backup register %u is not allowed to be written",
 					reg_idx);
@@ -431,8 +436,8 @@ static int *stm32_tamp_nvram_get_backup_
 	u32 offset_field;
 
 	idx_bkpreg_zones_end = devm_kcalloc(dev,
-					    sizeof(*idx_bkpreg_zones_end),
 					    nb_zones,
+					    sizeof(*idx_bkpreg_zones_end),
 					    GFP_KERNEL);
 	if (!idx_bkpreg_zones_end) {
 		dev_err(dev, "Can't allocate registers zones\n");
